File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int breakingPoint = -1;
        
        int n=nums.size();
        
        for(int i=n-2;i>=0;i--)
        {
            if(nums[i]<nums[i+1])
            {
                breakingPoint = i;
                cout<<"found breakingpoint "<<breakingPoint<<endl;
                break;
            }
        }
        if(breakingPoint==-1)
        {
            reverse(nums.begin(),nums.end());
            return;
        }
        int swapElement=breakingPoint;
        for(int i=n-1;i>breakingPoint;i--)
        {
            if(nums[i]>nums[breakingPoint] )
            {
                cout<<"found biggest ele "<<i<<endl;
        swap(nums[breakingPoint],nums[i]);
                break;
            }
        }
                        cout<<"swapping done" << endl;
    for(int i=0;i<n;i++)
        cout<<nums[i]<<" ";
                                cout<<endl<<"reverse done" << endl;

        reverse(nums.begin()+breakingPoint+1,nums.end());  
            for(int i=0;i<n;i++)
        cout<<nums[i]<<" ";
    }
};

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void reverseArr(int[] nums,int start,int end)
    {
        while(start<=end)
        {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
    
    public void nextPermutation(int[] nums) {
        int i;
        for(i=nums.length-2;i>=0;i--)
        {
            if(nums[i]<nums[i+1]) break;
        }
        if(i==-1) { reverseArr(nums,0,nums.length-1);return;
                  }
        int swapElement = i;
        for(int j=nums.length-1;j>i;j--)
        {
            if(nums[j] > nums[i]) {swapElement = j;break;}
        }
        int temp = nums[i];
        nums[i] = nums[swapElement];
        nums[swapElement] = temp;
        reverseArr(nums,i+1,nums.length-1);
        return;
    }
}